### PR TITLE
Deduplicate common snippets

### PR DESCRIPTION
**Pull request by tocic**
Extracts:

- `Question.objects.filter(state='PUB')`
- `session`, `UserData` and `register_correct_answer()`

Makes the `create_questions()` helper more flexible to use it in more contexts.

And makes an additional assertion in `test_clear_correct_answers` as requested.

### CODE REVIEW

File: quiz/test_auto_publish.py [hunk 1]
```diff
@@ -1,31 +1,35 @@
...
 
 from django.core.management import call_command
 from django.test import TestCase
 from django.utils import timezone
 
-from quiz.models import Question
+from quiz.test_helpers import create_questions
+
+
+def publish_and_refresh_scheduled_questions(questions):
```
Comment (knatten) [inline quiz/test_auto_publish.py:11]: This is only ever called with a single question. Would it be simpler to make it take a single question rather than a collection?

File: quiz/test_auto_publish.py [hunk 2]
```diff
@@ -1,31 +1,35 @@
...
+    call_command("auto_publish", "--skip-tweet")
+    for q in questions:
+        q.refresh_from_db()
```
Comment (knatten) [inline quiz/test_auto_publish.py:14]: My Django is very rusty. Why was this changed from `save` to `refresh_from_db`?

File: quiz/util.py [hunk 3]
```diff
@@ -6,3 +9,7 @@ def get_client_ip(request):
     else:
         ip = request.META.get('REMOTE_ADDR')
     return ip
+
+
+def get_published_questions():
```
Comment (knatten) [inline quiz/util.py:14]: Maybe we'll find a better home for this eventually, but it can live here for now until a natural group of functions emerges that we can extract to somewhere more meaningful than `util`.

File: quiz/test_session.py [hunk 4]
```diff
@@ -2,57 +2,56 @@
...
 
     def test_given_existing_session__correct_answers_are_still_there(self):
-        q = Question.objects.create(state='PUB', difficulty=1)
-        old_data = UserData({})
-        old_data.register_correct_answer(q.pk)
-        session = {'user_data': old_data}
-        new_data = UserData(session)
+        [q] = create_questions(1)
+        old_data = create_session_with_answers_to((q,))
+        new_data = UserData({'user_data': old_data})
```
Comment (knatten) [inline quiz/test_session.py:47]: I think I'd keep the `session` temporary variable, to show more clearly that we're simulating creating `UserData` from a session.

File: quiz/test_session.py [hunk 5]
```diff
@@ -2,57 +2,56 @@
...
         self.assertSetEqual({q.pk}, new_data.get_correctly_answered_questions())
 
     def test_clear_correct_answers(self):
-        q = Question.objects.create(state='PUB', difficulty=1)
-        s = UserData({})
-        s.register_correct_answer(q.pk)
-        s.clear_correct_answers()
-        self.assertSetEqual(set(), s.get_correctly_answered_questions())
+        [q] = create_questions(1)
+        user_data = create_session_with_answers_to((q,))
```
Comment (knatten) [inline quiz/test_session.py:52]: While at it, maybe add an assertion here that there is in fact at this point one correctly answered question?

File: quiz/test_auto_publish.py (hunk 1 above)
Comment (tocic) [reply quiz/test_auto_publish.py:11]: Yes, it would be simpler from the call site. But my reasoning was: 1) to support any number of questions if the need arises; 2) but mainly to match the function name: we could have scheduled multiple questions outside of this function, so passing it only one question doesn't necessary mean that we will be publishing only this question. The problem is that the function depends on the global state (we can't pass a list of questions to the `auto_publish` command, AFAIK), and even now, the name is a little misleading, since we publish only the `SCH` questions, but want to refresh every question from the passed iterable. So for the function with one parameter the correct name would be something like `publish_scheduled_questions_and_refresh_this_question`. For multiple parameters `publish_scheduled_questions_and_refresh_these_questions`. The latter has been roughly transformed to the current shorter version. Let me know if you still want to change the number of parameters or have a better name and I'll fix that.

File: quiz/util.py (hunk 3 above)
Comment (tocic) [reply quiz/util.py:14]: Yeah, I think it's fine for now. But can create `model_queries.py`, `query_filters.py`, or something similar if you wish.

**Comment by tocic** [comment]
@knatten I've replied to the review and also rebased. Check out [the diff](https://github.com/knatten/cppquiz/compare/0cfa7158eb841ab8a2f715cc40e7a30005ebf6b1..065fb18e5530ea8d6bbef18d031def11867cf4c4).

File: quiz/test_auto_publish.py (hunk 1 above)
Comment (knatten) [reply quiz/test_auto_publish.py:11]: Right, I didn't think about that. I actually think the tests are easier to understand if we just do this then:

```python
def auto_publish_all_questions():
    call_command("auto_publish", "--skip-tweet")
```

```python
def test_publishes_scheduled_question_from_the_past(self):
    past = timezone.now() - timedelta(hours=1)
    [q] = create_questions(1, state="SCH", publish_time=past)
    auto_publish_all_questions()
    q.refresh_from_db()
    self.assertEqual("PUB", q.state)
```

What do you think?

Btw I've noticed all your comments in the other issues too, I just have a bit limited time. I'm getting to them! :)

File: quiz/util.py (hunk 3 above)
Comment (knatten) [reply quiz/util.py:14]: Let's just extract it later if an obvious place appears.

File: quiz/test_auto_publish.py (hunk 1 above)
Comment (tocic) [reply quiz/test_auto_publish.py:11]: That fixes the name issue. The reason I haven't splitted those 2 actions is that it's easy to forget to call `q.refresh_from_db()` after seemingly unrelated to `q` function `auto_publish_all_questions()` and get a dangerous illusion of a passed test. Let's just leave this snippet as is. Anyway, `Question.objects.get()` is used in other places instead of `refresh_from_db()` and extracting a single `call_command()` doesn't gain much benefit.

File: quiz/test_session.py (hunk 5 above)
Comment (tocic) [reply quiz/test_session.py:53]: I've made a separate commit for that, as it's unrelated to the deduplication per se (perhaps worth a separate PR even). And, to make it easier to review and revert, also have separated other changes. Feel free to squash them if you want.

File: quiz/test_auto_publish.py (hunk 1 above)
Comment (knatten) [reply quiz/test_auto_publish.py]: Ok, let's just leave this.

File: quiz/test_session.py (hunk 5 above)
Comment (knatten) [reply quiz/test_session.py:53]: Nice! I appreciate the splitting into commits. I prefer leaving them as separate commits when merging the PR. Thanks a lot for the help!
